src/System.Diagnostics.Process/src/System/Diagnostics/Process.Unix.cs

```diff
@@ -63,6 +63,49 @@ public void Kill()
             }
         }
 
+        /// <summary>Stops the associated process immediately.</summary>
+        private void Stop()
+        {
+            EnsureState(State.HaveNonExitedId);
+            if (Interop.Sys.Kill(_processId, Interop.Sys.Signals.SIGSTOP) != 0)
+            {
+                throw new Win32Exception(); // to match behavior of Kill()
+            }
+        }
+
+        /// <summary>
+        /// Makes a best-effort attempt to kill all descendant (child, grandchild, etc.) processes.
+        /// </summary>
+        private void KillTree()
+        {
+            try
+            {
+                // Stop but don't kill the process. Keeps additional children from being started but leaves the process around 
+                // so that its children can be enumerated.
+                Stop();
+            }
+            catch
+            {
+                // Making a best attempt. If it fails (perhaps because the process is already dead), give up.
+                return;
+            }
+
+            IReadOnlyList<Process> children = GetChildProcesses();
```

Comment:
Each call to GetChildProcesses ends up enumerating all of the processes, which means this can end up being O(N*M).  Could we instead just get a list of processes once, and then operate on that fixed list?  I realize the processes involved may be creating additional processes while we're getting the list, so it's possible we could end up missing some.  Maybe that's a good enough reason not to do this, but I think it'd be worthwhile to understand how bad the costs could be.

Reply:
It's a choice between tradeoffs: The advantage of the recursive approach is that we can stop each descendant process from starting new processes before we enumerate its children. The downside is enumerating processes multiple times.

Do you think a simple BenchmarkDotNet from my workstation be sufficient for assessing the costs or is something fancier needed?